File: peao.cpp
/*
Projeto: Jogo de Xadrez
3º Semestre de Ciências da Computação - UFSCar Sorocaba
Programação orientada a objetos
Profª Katti Faceli
Integrantes: Giovanni Alvarenga, Fábio Yamaya, João Victor Vilar
Implementação da classe "peao"
Descrição:Implementa as funções da classe "peao" contidas no arquivo .h
*/
#include "peao.h"

//Construtor da classe
Peao::Peao(int _cor, char _tipo) : Peca(_cor, _tipo) {}

//Destrutor da classe
//Peao::~Peao(){}

bool Peao::checaMovimento(int linhaOrigem, int colunaOrigem, int linhaDestino, int colunaDestino, bool haPeca) {
    //Calcula a diferença entre linha e coluna para obter o movimento
    int diferencaLinha = linhaDestino - linhaOrigem;
    int diferencaColuna = colunaDestino - colunaOrigem;
    //cout << "função do peao" << endl;
    //Verifica a movimentação de acordo com a peça
    if (haPeca == 0) {
        if (jaMoveu) {
            if (cor == 1 && diferencaLinha == 1 && diferencaColuna == 0) {
                return true;
            } else if (cor == 0 && diferencaLinha == -1 && diferencaColuna == 0) {
                return true;
            }
        } else {
            jaMoveu = 1;
            if (cor == 1 && (diferencaLinha == 1 || diferencaLinha == 2) && diferencaColuna == 0) {
                return true;
            }
            if (cor == 0 && (diferencaLinha == -1 || diferencaLinha == -2) && diferencaColuna == 0) {
                return true;
            }
        }
        return false;
    } else {
        if (colunaOrigem == colunaDestino) {
            return false;
        }

        if (diferencaColuna == 1 || diferencaColuna == -1) {
            if (cor == 1 && diferencaLinha == 1) {
                return true;
            } else if (cor == 0 && diferencaLinha == -1) {
                return true;
            }
        }
        return false;
    }
}
/* verificacoes:
        - Preto sempre começa em cima
        -se a peça sai do tabuleiro
        - se ja existe uma peca na posicao desejada
        -se respeita as regras de movimento da peca
*/

//branco = 0 MAIUSCULO
//preto = 1 minusculo
void Peao::desenha() {
    if (cor == 1) {
        cout << "p";
    } else {
        cout << "P";
    }
}

File: Peao.java
public class Peao extends Peca{

    private boolean jaMoveu;

    public Peao(int _cor) {
        this.cor = _cor;
        this.estado = true;
        this.jaMoveu = false;
    }

    public boolean checaMovimento(int linhaOrigem, int colunaOrigem, int linhaDestino, int colunaDestino, boolean haPeca) {
        if ((linhaDestino == linhaOrigem) &&(colunaDestino == colunaOrigem)){
            return false;
        }
        
        //Verifica somente se o movimento da peça é valido, sem considerar os limetes do tabuleiro, que será verificado pela classe Tabuleiro
        int diferencaLinha = linhaDestino - linhaOrigem;
        int diferencaColuna = colunaDestino - colunaOrigem;
           //Caso a posição de destino esteja vazia
        if (haPeca == false) {
            //Caso o peao ja tenha se movido antes, não poderá mais executar o movimento de 2 casas
            if (jaMoveu) {
                if (this.cor == 1 && diferencaLinha == 1 && diferencaColuna == 0) {
                    System.out.println("Movimento valido para Peão: "+linhaOrigem+","+colunaOrigem+"->"+linhaDestino+","+colunaDestino);
                    return true;
                } else if (this.cor == 0 && diferencaLinha == -1 && diferencaColuna == 0) {
                    System.out.println("Movimento valido para Peão: "+linhaOrigem+","+colunaOrigem+"->"+linhaDestino+","+colunaDestino);
                    return true;
                }
            }
            //Caso o peão ainda não tenha se movido, poderá se mover 2 casas
            else {
                jaMoveu = true;
                if (this.cor == 1 && (diferencaLinha == 1 || diferencaLinha == 2) && diferencaColuna == 0) {
                    System.out.println("Movimento valido para Peão: "+linhaOrigem+","+colunaOrigem+"->"+linhaDestino+","+colunaDestino);
                    return true;
                }
                if (this.cor == 0 && (diferencaLinha == -1 || diferencaLinha == -2) && diferencaColuna == 0) {
                    System.out.println("Movimento valido para Peão: "+linhaOrigem+","+colunaOrigem+"->"+linhaDestino+","+colunaDestino);
                    return true;
                }
            }
            System.out.println("Movimento invalido para Peão: "+linhaOrigem+","+colunaOrigem+"->"+linhaDestino+","+colunaDestino);
            return false;
        }
        //Caso exista uma peça na posição de destino
        else {//se a peca estiver na frente do peao ele nao pode se mover
            if (colunaOrigem == colunaDestino) {
                System.out.println("Movimento invalido para Peão: "+linhaOrigem+","+colunaOrigem+"->"+linhaDestino+","+colunaDestino);
                return false;
            }
            //se a peca estiver na diagonal do peao, ele pode se mover
            if (diferencaColuna == 1 || diferencaColuna == -1) {
                if (this.cor == 1 && diferencaLinha == 1) {
                    jaMoveu = true;
                    System.out.println("Movimento valido para Peão: "+linhaOrigem+","+colunaOrigem+"->"+linhaDestino+","+colunaDestino);
                    return true;
                } else if (this.cor == 0 && diferencaLinha == -1) {
                    jaMoveu = true;
                    System.out.println("Movimento valido para Peão: "+linhaOrigem+","+colunaOrigem+"->"+linhaDestino+","+colunaDestino);
                    return true;
                }
            }
            System.out.println("Movimento invalido para Peão: "+linhaOrigem+","+colunaOrigem+"->"+linhaDestino+","+colunaDestino);
            return false;
        }
    }
    //Desenha o peao
    public void desenha(){
        if (cor == 1) {
            System.out.print("p");
        } else { 
            System.out.print("P");
        }
    }
    
}




